libsolidity/analysis/ReferencesResolver.cpp

```diff
@@ -241,11 +241,15 @@ void ReferencesResolver::endVisit(ArrayTypeName const& _typeName)
 	if (Expression const* length = _typeName.length())
 	{
 		TypePointer lengthTypeGeneric = length->annotation().type;
+		auto arrayType = dynamic_cast<ArrayType const*>(baseType.get());
+		bool isByteOrFixedSizeArray = (arrayType && arrayType->isByteArray()) || !baseType->isDynamicallySized();
 		if (!lengthTypeGeneric)
 			lengthTypeGeneric = ConstantEvaluator(m_errorReporter).evaluate(*length);
 		RationalNumberType const* lengthType = dynamic_cast<RationalNumberType const*>(lengthTypeGeneric.get());
 		if (!lengthType || !lengthType->mobileType())
 			fatalTypeError(length->location(), "Invalid array length, expected integer literal or constant expression.");
+		else if (lengthType->isZero() && isByteOrFixedSizeArray)
```

Comment:
Why did you add the special cases? I think this can be simplified to just `lengthType->isZero()`.

Reply:
Ah, of course. I missed that this will only be executed if the length is specified (which is obviously not the case for dynamically-sized arrays 😉 )